### PR TITLE
feat(csharp): adding retry configuration

### PR DESCRIPTION
**Pull request by Ramanth**
Adding retry configuration using the Polly library.

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

**Comment by wing328** [comment]
cc @mandrean (2017/08), @jimschubert (2017/09) ❤️ @frankyjuang (2019/09) @shibayan (2020/02)

File: modules/openapi-generator/src/main/resources/csharp-netcore/ApiClient.mustache [hunk 1]
```diff
@@ -408,7 +410,26 @@ namespace {{packageName}}.Client
...
+                response = (policyResult.Outcome == OutcomeType.Successful) ? policyResult.Result : new RestResponse<T>
+                {
+                    Request = req,
+                    ErrorException = policyResult.FinalException
+                };
+            }
+            else
+            {
+                response = client.Execute<T>(req);
+            }
```
Comment (shibayan) [inline modules/openapi-generator/src/main/resources/csharp-netcore/ApiClient.mustache:432]: The inability to customize the retry strategy is a concern, and I think a design that injects the retry strategy via IConfiguration would be better.

**Comment by cariarer** [comment]
Hi all, I would really like to use the generated .net core client for my customer. The main thing that stopping me is actually the lack of support for a circuit breaker, like Polly. That's why I'm commenting here. What I was wondering about is: Why is RestSharp used, instead of injecting a IHttpClientFactory? This way, you can customize your heart out and still provide some sensible defaults, if you want to use the client directly via simple constructor, providing a config or the endpoint url.

File: modules/openapi-generator/src/main/resources/csharp-netcore/ApiClient.mustache (hunk 1 above)
Comment (Ramanth) [reply modules/openapi-generator/src/main/resources/csharp-netcore/ApiClient.mustache:432]: @shibayan agree with you, thought of implementing that way itself, but the return type being a generic type IRestReponse<T> is what making it difficult to implement it via configuration.
Or least i couldn't think of a solution to implement that in a easier way. Any inputs along that lines will be helpful.

Comment (Ramanth) [reply modules/openapi-generator/src/main/resources/csharp-netcore/ApiClient.mustache:427]: @shibayan  Figured out a way to have the retry policy through configuration. As a workaround found that the RestSharp's ExecAsync< T > can be seperated in two steps ExecAsync and then Deserialize< T > which makes ExecAsync free from having generic type and thus i can define the retry policy free from generic and so able to inject through configuration..
Please take a look at it once...

**Comment by Ramanth** [comment]
while the chekcs are failing, i raised another PR with same changes just to test if something is wrong with this branch and the other [PR ](https://github.com/OpenAPITools/openapi-generator/pull/6281)checks are successful....so i guess the checks should be good.

Can everyone please take a look at the new changes and review them? that would be helpful !!
Thanks.

File: modules/openapi-generator/src/main/resources/csharp-netcore/ApiClient.mustache (hunk 1 above)
Comment (Ramanth) [reply modules/openapi-generator/src/main/resources/csharp-netcore/ApiClient.mustache:427]: @shibayan can you take a look again when you get time.

File: modules/openapi-generator/src/main/resources/csharp-netcore/Configuration.mustache [hunk 2]
```diff
@@ -10,6 +10,10 @@ using System.IO;
 using System.Linq;
 using System.Security.Cryptography.X509Certificates;
 using System.Text;
+using System.Net;
+using Polly;
+using Polly.Retry;
```
Comment (jimschubert) [inline modules/openapi-generator/src/main/resources/csharp-netcore/Configuration.mustache:15]: This configuration object is meant to be framework agnostic.

By adding RestSharp and Polly here, this breaks our contract of allowing users to build without RestSharp as a dependency and now adds Polly into the mix.

If the decision here is to tightly couple Polly with the RestSharp implementation of ApiClient, please modify only the ApiClient (preferably via constructor or thread-safe property injection).

File: modules/openapi-generator/src/main/resources/csharp-netcore/IReadableConfiguration.mustache [hunk 3]
```diff
@@ -2,7 +2,11 @@
 
 using System;
 using System.Collections.Generic;
+using System.Net;
 using System.Security.Cryptography.X509Certificates;
+using Polly;
+using Polly.Retry;
```
Comment (jimschubert) [inline modules/openapi-generator/src/main/resources/csharp-netcore/IReadableConfiguration.mustache:8]: We can't mix non-System types into this interface.

**Comment by wing328** [comment]
@Ramanth can you please resolve the merge conflicts when you've time? Then I'll merge. Thank you.

**Comment by wing328** [comment]
```
To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
dotnet build samples\client\petstore\csharp-netcore\OpenAPIClient\Org.OpenAPITools.sln
Microsoft (R) Build Engine version 16.6.0+5ff7b0c9e for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.
  Determining projects to restore...
  Restored C:\projects\openapi-generator\samples\client\petstore\csharp-netcore\OpenAPIClient\src\Org.OpenAPITools\Org.OpenAPITools.csproj (in 1.47 sec).
  Restored C:\projects\openapi-generator\samples\client\petstore\csharp-netcore\OpenAPIClient\src\Org.OpenAPITools.Test\Org.OpenAPITools.Test.csproj (in 4.62 sec).
Client\ApiClient.cs(529,78): error CS0029: Cannot implicitly convert type 'System.Runtime.CompilerServices.ConfiguredTaskAwaitable<RestSharp.IRestResponse<T>>' to 'System.Threading.Tasks.Task<RestSharp.IRestResponse>' [C:\projects\openapi-generator\samples\client\petstore\csharp-netcore\OpenAPIClient\src\Org.OpenAPITools\Org.OpenAPITools.csproj]
Client\ApiClient.cs(529,78): error CS1662: Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type [C:\projects\openapi-generator\samples\client\petstore\csharp-netcore\OpenAPIClient\src\Org.OpenAPITools\Org.OpenAPITools.csproj]
Client\ApiClient.cs(538,29): error CS0029: Cannot implicitly convert type 'System.Runtime.CompilerServices.ConfiguredTaskAwaitable<RestSharp.IRestResponse<T>>' to 'RestSharp.IRestResponse<T>' [C:\projects\openapi-generator\samples\client\petstore\csharp-netcore\OpenAPIClient\src\Org.OpenAPITools\Org.OpenAPITools.csproj]
Build FAILED.
Client\ApiClient.cs(529,78): error CS0029: Cannot implicitly convert type 'System.Runtime.CompilerServices.ConfiguredTaskAwaitable<RestSharp.IRestResponse<T>>' to 'System.Threading.Tasks.Task<RestSharp.IRestResponse>' [C:\projects\openapi-generator\samples\client\petstore\csharp-netcore\OpenAPIClient\src\Org.OpenAPITools\Org.OpenAPITools.csproj]
Client\ApiClient.cs(529,78): error CS1662: Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type [C:\projects\openapi-generator\samples\client\petstore\csharp-netcore\OpenAPIClient\src\Org.OpenAPITools\Org.OpenAPITools.csproj]
Client\ApiClient.cs(538,29): error CS0029: Cannot implicitly convert type 'System.Runtime.CompilerServices.ConfiguredTaskAwaitable<RestSharp.IRestResponse<T>>' to 'RestSharp.IRestResponse<T>' [C:\projects\openapi-generator\samples\client\petstore\csharp-netcore\OpenAPIClient\src\Org.OpenAPITools\Org.OpenAPITools.csproj]
    0 Warning(s)
    3 Error(s)

```
Appveyor CI failed with the above error message. Please take a look when you've time.

**Pull request merged**